Mathlib/GroupTheory/Congruence.lean

```diff
@@ -1147,10 +1147,42 @@ instance one [MulOneClass M] (c : Con M) : One c.Quotient where
   one := Quotient.mk'' (1 : M)
   -- one := ((1 : M) : c.Quotient)
 
+/-- The proposition that the `+ᵥ` operation is compatible with the congruence relation.
+
+Note that this is similar to `AddAction.QuotientAction`, but with slightly different assumptions. -/
+class _root_.AddCon.CompatibleVAdd (α : Type*) [Add M] [VAdd α M] (c : AddCon M) :
+    Prop where
+  rel_smul (a : α) {w x : M} : c w x → c (a +ᵥ w) (a +ᵥ x)
+
+/-- The proposition that the `•` operation is compatible with the congruence relation.
+
+Note that this is similar to `MulAction.QuotientAction`, but with slightly different assumptions. -/
+@[to_additive]
+class CompatibleSMul (α : Type*) [Mul M] [SMul α M] (c : Con M) : Prop where
+  rel_smul (a : α) {w x : M} : c w x → c (a • w) (a • x)
+
+/-- This instance notably works for `α = M`. -/
+@[to_additive]
+instance CompatibleSMul.ofIsScalarTower {α M : Type*}
+    [MulOneClass M] [SMul α M] [IsScalarTower α M M] (c : Con M) : CompatibleSMul α c where
+  rel_smul a w x h := by simpa only [smul_one_mul] using c.mul (c.refl' (a • (1 : M) : M)) h
+
+/-- This instance notably works for `α = Mᵐᵒᵖ`. -/
 @[to_additive]
-theorem smul {α M : Type*} [MulOneClass M] [SMul α M] [IsScalarTower α M M] (c : Con M) (a : α)
-    {w x : M} (h : c w x) : c (a • w) (a • x) := by
-  simpa only [smul_one_mul] using c.mul (c.refl' (a • (1 : M) : M)) h
+instance CompatibleSMul.ofSMulCommClass {α M : Type*}
+    [MulOneClass M] [SMul α M] [SMulCommClass α M M] (c : Con M) : CompatibleSMul α c where
+  rel_smul a w x h := by simpa only [mul_smul_one] using c.mul h (c.refl' (a • (1 : M) : M))
+
+/-- This instance notably works for `α = Mᵐᵒᵖ`. -/
+@[to_additive]
+instance CompatibleSMul.ofSMulCommClass' {α M : Type*}
+    [MulOneClass M] [SMul α M] [SMulCommClass M α M] (c : Con M) : CompatibleSMul α c :=
+  letI := SMulCommClass.symm; inferInstance
```

Comment:
Out of scope for this PR, but can we still not make `SMulCommClass α β γ → SMulCommClass β α γ` an instance? I thought Lean 4 was better (apparently not perfect!) at dealing with instance loops.